crates/pallet-humanode-offences/src/lib.rs

```diff
@@ -0,0 +1,94 @@
+
+use frame_support::{
+    sp_std::prelude::*,
+    traits::{Get, StorageVersion},
+};
+pub use pallet::*;
+use sp_staking::offence::{Offence, OffenceError, ReportOffence};
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+/// The current storage version.
+const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+// We have to temporarily allow some clippy lints. Later on we'll send patches to substrate to
+// fix them at their end.
+#[allow(clippy::missing_docs_in_private_items)]
+#[frame_support::pallet]
+pub mod pallet {
+    use frame_support::pallet_prelude::*;
+
+    use super::*;
+
+    /// Configure the pallet by specifying the parameters and types on which it depends.
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_humanode_session::Config {
+        /// Deauthentication reason on offence report.
+        type DeauthenticationReasonOnOffenceReport: Get<Self::DeauthenticationReason>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    pub struct Pallet<T>(_);
+
+    /// The total number of offences.
+    #[pallet::storage]
+    pub type Total<T: Config> = StorageValue<_, u64>;
+}
+
+/// The offender type alias.
+pub(crate) type OffenderOf<T> = pallet_humanode_session::IdentificationTupleFor<T>;
+
+impl<T: Config, O> ReportOffence<T::AccountId, OffenderOf<T>, O> for Pallet<T>
+where
+    O: Offence<OffenderOf<T>>,
+{
+    fn report_offence(_reporters: Vec<T::AccountId>, offence: O) -> Result<(), OffenceError> {
+        let offenders = offence.offenders();
+        let mut should_be_deauthenticated = Vec::with_capacity(offenders.len());
+
+        for offender in offenders {
+            let (_offender, identity) = &offender;
+            match identity {
+                pallet_humanode_session::Identification::Bioauth(authentication) => {
+                    should_be_deauthenticated.push(authentication.public_key.clone());
+                }
+                pallet_humanode_session::Identification::Bootnode(..) => {
+                    // Never slash the bootnodes.
+                }
+            }
+        }
+
+        if !should_be_deauthenticated.is_empty() {
+            let offences_number: u64 = should_be_deauthenticated
+                .len()
+                .try_into()
+                .expect("u64 is big enough for this overflow to be practically impossible");
+
+            let _ = <pallet_bioauth::Pallet<T>>::deauthenticate(
+                should_be_deauthenticated,
+                T::DeauthenticationReasonOnOffenceReport::get(),
+            );
+
+            <Total<T>>::mutate(|total| {
+                let new_total = total.map_or(offences_number, |t| {
+                    t.checked_add(offences_number)
+                        .expect("u64 is big enough for this overflow to be practically impossible")
+                });
+                *total = Some(new_total);
+            });
+        }
+
+        Ok(())
```

Comment:
`pallet_offences / report_offence` deposits `Event::offence` — it's useless for us 'cause we have `pallet_bioauth / Event::AuthRemoved` ? I'd like to note that it's slightly less informative. This comment is just a reminder to discuss events and expected reasons to emit them.

Reply:
Good thoughts, I was thinking about it as well. But forgot to implement it. Thanks!

The general logic can be extended in theory (not just deauthentication) and it's gonna be good to explicitly submit events to the chain about offences if I would like to subscribe just on offence logic for example.

Reply:
https://github.com/humanode-network/humanode/pull/1291/commits/6ef1117c2a975cbd370ab29048881c9cbb3d705b